bindings/pydeck-carto/pydeck_carto/carto_auth/oauth2.py

```diff
@@ -0,0 +1,296 @@
+import os
+import sys
+import webbrowser
+from http.server import HTTPServer, BaseHTTPRequestHandler
+from urllib.parse import parse_qsl
+import requests
+import logging
+from urllib.parse import urlparse, urlencode
+
+from pydeck_carto.carto_auth.errors import CredentialsError
+
+logger = logging.getLogger(__name__)
+
+
+def get_host_port(netloc):
+    if ":" in netloc:
+        host, port = netloc.split(":", 1)
+        port = int(port)
+    else:
+        host = netloc
+        port = None
+
+    return host, port
+
+
+class CartoPKCE:
+    """Implements PKCE Authorization Flow for client apps"""
+
+    OAUTH_AUTHORIZE_URL = "https://auth.carto.com/authorize"
+    OAUTH_TOKEN_URL = "https://auth.carto.com/oauth/token"
+    CLIENT_ID = "AtxvHDeuXlR8XPfF2nj2Uv2I29pvmCxu"
+    REDIRECT_URI = "http://localhost:10000/callback/"
+    REDIRECT_URI_CLI = "https://callbacks.carto.com/token"
+
+    def __init__(
+        self,
+        state=None,
+        proxies=None,
+        open_browser=True,
+        requests_session=None,
+        requests_timeout=None,
+    ):
+        """
+        Creates PKCE Auth flow.
+
+        Parameters:
+             * state: Optional
+             * proxies: Optional, proxy for the Requests library to route through
+             * requests_timeout: Optional, tell Requests to stop waiting for a
+                                 response after a given number of seconds
+             * open_browser: Optional, whether the web browser should be opened to
+                             authorize a user
+        """
+        using_google_colab = 'google.colab' in sys.modules
+        if using_google_colab and open_browser:
```

Comment:
Set open_browser False if using_google_colab, instead of throwing (always) an exception.
```